src/meltano/core/hub/client.py

```diff
@@ -0,0 +1,252 @@
+"""Meltano Hub Client."""
+
+from __future__ import annotations
+
+from typing import Any
+
+import requests
+from structlog.stdlib import get_logger
+
+import meltano
+from meltano.core.plugin import (
+    BasePlugin,
+    PluginDefinition,
+    PluginRef,
+    PluginType,
+    Variant,
+)
+from meltano.core.plugin.error import PluginNotFoundError
+from meltano.core.plugin.factory import base_plugin_factory
+from meltano.core.plugin_discovery_service import PluginRepository
+from meltano.core.project import Project
+
+from .schema import IndexedPlugin, VariantRef
+
+logger = get_logger(__name__)
+
+
+class HubPluginTypeNotFound(Exception):
+    """Raised when a Hub plugin type is not found."""
+
+    def __init__(self, plugin_type: PluginType):
+        """Create a new HubPluginVariantNotFound.
+
+        Args:
+            plugin_type: The type of the plugin.
+        """
+        self.plugin_type = plugin_type
+
+    def __str__(self) -> str:
+        """Return a string representation of the error.
+
+        Returns:
+            The string representation of the error.
+        """
+        return "{type} is not supported in Meltano Hub. Available plugin types: {types}".format(
+            type=self.plugin_type.descriptor.capitalize(),
+            types=list(PluginType),
+        )
+
+
+class HubPluginVariantNotFound(Exception):
+    """Raised when a Hub plugin variant is not found."""
+
+    def __init__(
+        self,
+        plugin_type: PluginType,
+        plugin: IndexedPlugin,
+        variant_name: str,
+    ):
+        """Create a new HubPluginVariantNotFound.
+
+        Args:
+            plugin_type: The type of the plugin.
+            plugin: The indexed plugin.
+            variant_name: The name of the variant that was not found.
+        """
+        self.plugin_type = plugin_type
+        self.plugin = plugin
+        self.variant_name = variant_name
+
+    def __str__(self) -> str:
+        """Return a string representation of the error.
+
+        Returns:
+            The string representation of the error.
+        """
+        return "{type} '{name}' variant '{variant}' is not known to Meltano. Variants: {variant_labels}".format(
+            type=self.plugin_type.descriptor.capitalize(),
+            name=self.plugin.name,
+            variant=self.variant_name,
+            variant_labels=self.plugin.variant_labels,
+        )
+
+
+class MeltanoHubService(PluginRepository):
+    """PluginRepository implementation for the Meltano Hub."""
+
+    BASE_URL = "https://hub.meltano.com/meltano/api/v1"
```

Comment:
We should probably make this configurable, at least for testing purposes.

It would be nice to be able to stand up the Hub API from its main branch (or a feature one) and test Meltano against it directly instead of mocking it entirely.